llvm/docs/Telemetry.rst

```diff
@@ -70,120 +69,168 @@ Key components
+  class JsonSerializer : public Serializer {
+  public:
+    json::Object *getOutputObject() { return object.get(); }
+
+    llvm::Error start() override {
+      if (started)
+        return createStringError("Serializer already in use");
+      started = true;
+      object = std::make_unique<json::Object>();
+      return Error::success();
+    }
+
+    void writeBool(StringRef KeyName, bool Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void writeInt32(StringRef KeyName, int Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void writeSizeT(StringRef KeyName, size_t Value) override {
+      writeHelper(KeyName, Value);
+    }
+    void writeString(StringRef KeyName, StringRef Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void writeKeyValueMap(StringRef KeyName,
+                          std::map<std::string, std::string> Value) override {
+      json::Object Inner;
+      for (auto kv : Value) {
+        Inner.try_emplace(kv.first, kv.second);
       }
-    };
-    
-    // This defines a custom TelemetryInfo that has an addition Msg field.
-    struct MyTelemetryInfo : public llvm::telemetry::TelemetryInfo {
-      std::string Msg;
+      writeHelper(KeyName, json::Value(std::move(Inner)));
+    }
+
+    llvm::Error finish() override {
+      if (!started)
+        return createStringError("Serializer not currently in use");
+      started = false;
+      return Error::success();
+    }
+
+  private:
+    template <typename T> void writeHelper(StringRef Name, T Value) {
+      assert(started && "serializer not started");
+      object->try_emplace(Name, Value);
+    }
+    bool started = false;
+    std::unique_ptr<json::Object> object;
+  };
+       
+  class MyManager : public telemery::Manager {
+  public:
+  static std::unique_ptr<MyManager> createInstatnce(telemetry::Config* config) {
+    // If Telemetry is not enabled, then just return null;
+    if (!config->EnableTelemetry) return nullptr;
+
+    return std::make_unique<MyManager>();
+  }
+  MyManager() = default;
+
+  Error dispatch(TelemetryInfo* Entry) const override {
+    Entry->SessionId = SessionId;
+    emitToAllDestinations(Entry);
+  }
       
-      json::Object serializeToJson() const {
-        json::Object Ret = TelemeteryInfo::serializeToJson();
-        Ret.emplace_back("MyMsg", Msg);
-        return std::move(Ret);
+  void addDestination(std::unique_ptr<Destination> dest) override {
+    destinations.push_back(std::move(dest));
+  }
+  
+  // You can also define additional instrumentation points.
+  void logStartup(TelemetryInfo* Entry) {
+    // Add some additional data to entry.
+    Entry->Msg = "Some message";
+    dispatch(Entry);
+  }
+  
+  void logAdditionalPoint(TelemetryInfo* Entry) {
+    // .... code here
+  }
+  
+  private:
+    void emitToAllDestinations(const TelemetryInfo* Entry) {
+      for (Destination* Dest : Destinations)
+        Dest->receiveEntry(Entry);
       }
```

Comment:
GitHub unhelpfully highlighted the wrong bracket, making me think there was one missing...

In general, it may be worth locally putting all this sample code in a source file, then running the compiler and clang-format on it, so that there are no fundamental errors and the style is consistent with LLVM (there are plenty of minor formatting issues that I haven't commented on). You can put it back in the docs again after verifying things.

Reply:
done